internal/bminventory/inventory_v2_handlers.go

```diff
@@ -451,3 +454,95 @@ func (b *bareMetalInventory) V2GetPresignedForClusterCredentials(ctx context.Con
 	}
 	return installer.NewV2GetPresignedForClusterCredentialsOK().WithPayload(&models.Presigned{URL: &url})
 }
+
+func (b *bareMetalInventory) GetInfraEnvDownloadURL(ctx context.Context, params installer.GetInfraEnvDownloadURLParams) middleware.Responder {
+	if b.ImageServiceBaseURL == "" {
+		return common.NewApiError(http.StatusConflict, fmt.Errorf("missing image serviec base URL"))
+	}
+
+	infraEnv, err := common.GetInfraEnvFromDB(b.db, params.InfraEnvID)
+	if err != nil {
+		return common.GenerateErrorResponder(err)
+	}
+
+	osImage, err := b.getOsImageOrLatest(&infraEnv.OpenshiftVersion, infraEnv.CPUArchitecture)
+	if err != nil {
+		return common.GenerateErrorResponder(err)
+	}
+	if osImage.OpenshiftVersion == nil {
+		return common.GenerateErrorResponder(errors.Errorf("OS image entry '%+v' missing OpenshiftVersion field", osImage))
+	}
+
+	newURL, err := b.generateImageDownloadURL(infraEnv.ID.String(), string(*infraEnv.Type), *osImage.OpenshiftVersion, infraEnv.CPUArchitecture, infraEnv.ImageTokenKey)
+	if err != nil {
+		return common.GenerateErrorResponder(err)
+	}
+
+	if err = b.db.Model(infraEnv).Update("download_url", newURL).Error; err != nil {
+		b.log.WithError(err).Errorf("Failed to update download_url for infraEnv %s", params.InfraEnvID)
+		return common.GenerateErrorResponder(err)
+	}
+
+	payload := models.InfraEnvImageURL{URL: newURL}
+
+	// parse the exp claim out of the url to return with the payload
+	parsedURL, err := url.Parse(newURL)
+	if err != nil {
+		return common.GenerateErrorResponder(err)
+	}
+
+	if tokenString := parsedURL.Query().Get("image_token"); tokenString != "" {
+		// we just created these claims so they are safe to parse unverified
+		token, _, err := new(jwt.Parser).ParseUnverified(tokenString, jwt.MapClaims{})
+		if err != nil {
+			return common.GenerateErrorResponder(err)
+		}
+		claims, ok := token.Claims.(jwt.MapClaims)
+		if !ok {
+			return common.GenerateErrorResponder(errors.Errorf("malformed token claims in url"))
+		}
+		exp, ok := claims["exp"].(float64)
+		if !ok {
+			return common.GenerateErrorResponder(errors.Errorf("token missing 'exp' claim"))
+		}
+		payload.Exp = fmt.Sprintf("%.f", exp)
+	}
```

Comment:
Shouldn't you update InfraEnv.expires_at?  Maybe move the logic into generateImageDownloadURL() or a separate method for reuse?

Reply:
I actually didn't notice that field until you mentioned it. Yeah I can match that with the currently set url's expiration.